libflux: add flux_future_reset()

garlick opened the pull request:
This PR adds `flux_future_reset()`, to "unfulfill" a future.

The driving use case was RPC with multiple response messages.  One can now do this:
```c
flux_future_t *f = flux_rpc (...)
do {
    flux_rpc_get (f, ...);
    flux_future_reset (f);
} while (condition);
flux_future_destroy (f);
```
or in a continuation, one can conditionally call either `flux_future_reset()` or `flux_future_destroy()` before returning.  If reset, the continuation will be called on the next fulfillment (RPC response, or whatever).

Added unit tests and updated man pages.  I reworked the "user" man page a bit for better readability also.

I'd say test coverage might be a little light here, so I'll see how that goes and possibly add more tests.

grondo commented:
Coverage looks good to me. Only case that doesn't appeared covered is when the timer_watcher needs to be reset instead of created in `then_context_set_timeout`:

https://codecov.io/gh/flux-framework/flux-core/pull/1503/diff#D1-194

This actually reminds me of a possible enhancement for `flux_future_reset` (or maybe this is possible already, I haven't checked). It would make the code cleaner for some multi-response use cases if the timer and/or continuation handler could be reset at the same time the future is reset. This would allow a chain of continuations, possibly with different timeouts, to be used with a multi-response RPC when that makes sense (e.g. continuation 1 creates some state, resets the future to call continuation 2 on next response which uses that state and so on)

Of course this could be done now by keeping a state variable in `void *` arg to the future, but I think code would be easier to follow in some cases if this wasn't required.

Perhaps this could be a new function `flux_future_again` with the same prototype as `flux_future_then` (or I guess `flux_future_then` called a fulfilled future could reset it? That could be bad design though..)

garlick commented:
I'll add a unit test for calling `flux_future_then()` again.  As an aside:  It shouldn't actually matter what order `flux_future_reset()` and `flux_future_then()` are called.

As I think we discussed, `flux_future_then()` cannot do an implicit reset, since some futures might be fulfilled immediately (say if something is already present in a local cache), and then the reset would cause that "event" to be ignored.

`flux_future_again()` would be OK, though I wonder if it's that useful given that `flux_future_reset()` shouldn't require error checking.  We could always add it later.

grondo commented:
> flux_future_again() would be OK, though I wonder if it's that useful given that flux_future_reset() shouldn't require error checking. We could always add it later.

Yeah, sorry, I came to the same conclusion in my follow up comment.

> As an aside: It shouldn't actually matter what order flux_future_reset() and flux_future_then() are called.

Ah good point. I was thinking maybe if the last future result was an error, this code would be triggered if `flux_future_reset` was not called before `flux_future_then`:
```C
   if (f->result_errnum_valid) {
        errno = f->result_errnum;
        return -1;
    }
```

garlick commented:
Oh indeed you are right about that case of error fulfillment impacting the required order.  Thanks!

garlick commented:
I fixed a bug, fixed a test bug (that was masking the other bug), and added coverage for calling `flux_future_then()` after  `flux_future_reset()`.  The order does matter, and I updated the man page to indicate this.

Then I squashed down the incremental development.

grondo commented:
Great! can this go in?
